From 51a3bf3aef4acd6b960c72b0ea039d76eecd013e Mon Sep 17 00:00:00 2001
From: Stefan Agner <stefan@agner.ch>
Date: Tue, 3 Apr 2018 13:54:13 +0200
Subject: [PATCH] tool_operate: resume at last byte when retrying

If continue/resume is enabled, try to also resume when retrying due to
transient connection problems.

Closes #1084

See also:
https://github.com/curl/curl/pull/2450

- Forward port to curl 7.69.1
- No longer throw away the already downloaded data when resuming with a
  fall back to not trying to resume if server doesn't support resuming.
  Taken partly from https://github.com/curl/curl/pull/2450

- Forward port to curl 7.87.0, remove fuzz
Signed-off-by: Max Krummenacher <max.krummenacher@toradex.com>
---
 src/tool_operate.c | 15 ++++++++++++++-
 1 file changed, 14 insertions(+), 1 deletion(-)

diff --git a/src/tool_operate.c b/src/tool_operate.c
index 79db063..5b42fcd 100644
--- a/src/tool_operate.c
+++ b/src/tool_operate.c
@@ -515,6 +515,11 @@ static CURLcode post_per_transfer(struct GlobalConfig *global,
         }
       }
     } /* if CURLE_OK */
+    else if(result == CURLE_RANGE_ERROR) {
+      /* Retry failed last time, don't retry with resume anymore */
+      config->use_resume = 0;
+      retry = RETRY_HTTP;
+    }
     else if(result) {
       const char *scheme;
 
@@ -577,13 +582,21 @@ static CURLcode post_per_transfer(struct GlobalConfig *global,
             "%ld retries left.\n",
             m[retry], sleeptime/1000L, per->retry_numretries);
 
+      /* If resume is requested, resume at last byte on retry too */
+      if (config->use_resume) {
+        fprintf(global->errors, "Continuing at %"
+                CURL_FORMAT_CURL_OFF_T " bytes\n",
+                outs->bytes);
+        my_setopt(curl, CURLOPT_RESUME_FROM_LARGE, outs->bytes);
+      }
+
       per->retry_numretries--;
       if(!config->retry_delay) {
         per->retry_sleep *= 2;
         if(per->retry_sleep > RETRY_SLEEP_MAX)
           per->retry_sleep = RETRY_SLEEP_MAX;
       }
-      if(outs->bytes && outs->filename && outs->stream) {
+      if(outs->bytes && outs->filename && outs->stream && !config->use_resume) {
         int rc;
         /* We have written data to an output file, we truncate file
          */
-- 
2.35.3

